### ISSUE TITLE
CronJob daemonset (previously ScheduledJob)

### ISSUE THREAD

**Issue by rothgar** [kubernetes/kubernetes#36601]
Feature Request:
  - Ability to run a scheduled job as a daemonset so it can run on each node. e.g. image garbage collection with something like [docker-gc](https://github.com/spotify/docker-gc/)

**Kubernetes version** (use `kubectl version`):
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.4", GitCommit:"3b417cc4ccd1b8f38ff9ec96bb50a81ca0ea9d56", GitTreeState:"clean", BuildDate:"2016-10-21T02:48:38Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.4", GitCommit:"3b417cc4ccd1b8f38ff9ec96bb50a81ca0ea9d56", GitTreeState:"clean", BuildDate:"2016-10-21T02:42:39Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}

**What you expected to happen**:
The ability to target more than one node in a scheduled job. If I specify a nodeSelector for workers the job will only run on 1 worker node. I would like the job to run on all worker nodes.

**How to reproduce it** (as minimally and precisely as possible):
```
apiVersion: batch/v2alpha1
kind: ScheduledJob
metadata:
  name: hello
spec:
  schedule: 0/1 * * * ?
  jobTemplate:
    spec:
      template:
        spec:
          containers:
          - name: hello
            image: busybox
            args:
            - /bin/sh
            - -c
            - date; echo Hello from the Kubernetes cluster
          restartPolicy: OnFailure
        nodeSelector:
          nodetype: worker
```

I think there is a notion of a node service being worked on but I A) couldn't find the link/doc describing how it works B) don't think it applies to scheduled jobs

**Comment by bgrant0607** [comment]
Would be possible if ScheduledJob could launch a Template.

cc @erictune @pwittrock

**Comment by kargakis** [comment]
cc: @soltysh

**Comment by soltysh** [comment]
> Would be possible if ScheduledJob could launch a Template.

Yup, I don't see any problems with CJ creating any type of resources available in the cluster. 

First I'd like to address the most important issues, so that we could graduate CJ to beta (maybe in 1.6) and then while in beta add features as this, @erictune agreed?

**Comment by bgrant0607** [comment]
This could also just be a DaemonSet of containers that sleep to create whatever duty cycle they want.

**Comment by rothgar** [comment]
@bgrant0607 That was actually the route I was going. I had a shell script that would just sleep for a day and then run the commands I wanted. I figured this should be supported natively. Hence this open issue.

**Comment by sheerun** [comment]
In my case kubernetes doesn't clear exited containers properly, and I need to run cleanup job on all workers periodically. This would be a useful feature. As for now probably the only solution is to setup cron job on each node...

**Comment by andrewwebber** [comment]
For me this is a very important issue but more in the broader scope of Jobs (not just scheduled).
CoreOS has deprecated Fleet (distributed systemd). With Fleet one could broadcast a global systemd unit to run on and update all machines in a cluster with a label selector.

Given a kubernetes cluster provisioned with no configuration management tool such as Puppet, Fleet or Chef it would be logical for kubernetes to support modification of its self using just kubernetes. This is why I personally would like job scheduling choice:
- Run or schedule a job on all machines
- Run or schedule a job on a sub-set of machines using a label selector
- Run or schedule a job on a single machine

**Comment by kargakis** [comment]
Maybe CronJobs could have a node selector and if specified, empty would mean run this on every node, otherwise select the nodes you want to run on. otherwise if unspecified run this as currently.

**Comment by soltysh** [comment]
Although I really like the NodeSelector idea, we need to be very careful about it. We currently struggling with explaining RestartPolicy which applies to a Pod not a Job itself. 

Having said that I'm thinking, maybe we should have a similar primitive for running Jobs on all nodes, like we have Deployments and DaemonSets. Eventually, I'm hoping (#28349) CronJobs will allow scheduling any kind of objects, so that part will be the least problem. @kubernetes/sig-apps-feature-requests @erictune wdyt?

**Comment by kow3ns** [comment]
How will CronJob interact with Nodes that are added to or removed from the cluster during a scheduled execution?

1. If a node is removed from the cluster during execution does that constitute a failure for that scheduled execution of the CronJob? 
2. If a node is added (say via auto-scaling) during execution of the current iteration of the CronJob, should the currently running CronJob schedule a new Pod on that node as well, or will the CronJob only run on that node during the next scheduled execution?

What are the guarantees, if any, that we will provide with respect to execution under these conditions?

**Comment by soltysh** [comment]
@kow3ns good points, I'm guessing that would depend on the action performed and should be configurable by a user. Having a strict requirement to go with one or the other options isn't good way, imo.

**Comment by everett-toews** [comment]
Here's my workaround until this feature become a reality, [Run Once DaemonSet on Kubernetes](http://blog.phymata.com/2017/05/13/run-once-daemonset-on-kubernetes/).

**Comment by kargakis** [comment]
Older issue asking about this: https://github.com/kubernetes/kubernetes/issues/17182

**Comment by taherv** [comment]
Thanks for the workaround @everett-toews, but some work needs to be done to ensure that workaround actually works when nodes are added to a cluster :(

Looking at it another way, and I'm sure we must have thought about this, why don't we use "exit code of 0" from a daemonset to mean "please dont restart" ? This approach has several advantages :
1. Use plain old DaemonSet object (no customizations necessary)
2. Avoid unnecessary sleep-forever pods (and dirty scripting just to keep something alive, for no reason).

The open question then would be : What is the status of the DaemonSet if the pod has exited successfully ? Is there a "Completed" status ? I guess we could use the same status as a completed job ? 

Thoughts ?

**Comment by kargakis** [comment]
DaemonSets are designed for long-running daemons, not for batch workloads. I still prefer we solve this in the CronJob API.

**Comment by rootfs** [comment]
I contribute a use case for run-once daemonset (details at https://github.com/ceph/ceph-docker/pull/733). 

When installing Ceph OSD daemonset on a node, we need to properly initialize the devices (make filesystem and get keyrings see `ceph-disk prepare`) before using them (i.e. `ceph-disk activate`). 

So we need two DS here: `ceph-osd-prepare` and `ceph-osd-activate`, where `ceph-osd-prepare` is a run-once DS, `ceph-osd-activate` is a long running DS.

**Comment by soltysh** [comment]
> I contribute a use case for run-once daemonset (details at ceph/ceph-docker#733).

This sounds more like a job daemonset, rather than cronjob one.

**Comment by kow3ns** [comment]
@rootfs Why can't you prepare and activate the osd storage using init containers on the DaemonSet?

**Comment by rootfs** [comment]
@kow3ns good point, let me try that, thanks for the pointer.

**Comment by resouer** [comment]
> DaemonSets are designed for long-running daemons, not for batch workloads. I still prefer we solve this in the CronJob API.

@kargakis Thanks for redirect. Would like to continue discussion here. Covering my use case (run-once job on every node) by CronJob seems make it more complicated.

1. How CronJob API  fits to the feature of "spreading to all Nodes"? Let's say, we can add a "every node" as we discussed.
2. But the Job daemon (run-once job) has no requirement for "Cron", ` schedule: 0/1 * * * ?` will not make sense in this case.
3. Or, we are actually mixing two issues here: CronJob which can run on every node, Job which can run on every node. That remind me **can we extract the `run on every node` logic out**, so both `Daemon`, `CronJob` and `Job` can share it in common? The both issues can be fixed in elegant way.

cc @kow3ns

**Comment by iMartyn** [comment]
Another thing to think about if using CronJobs is timings - If you want something running at intervals on all your nodes, you probably don't want them running on the same minute (stampede)  this would be a case for something like a ~ operator, for instance :
```
~5 * * * *
```
could run at close to :05 every hour, but if determined randomly by the node, (whilst possible, not likely) should not run on all nodes at the same time.

This would be useful in cronjobs in general not just run-once cronjobs.

**Comment by bgrant0607** [comment]
A once-per-node cronjob would be non-trivial to make it work reliably, since pods won't necessarily be able to schedule on all selected nodes. Theoretically that problem  exists for Jobs already, but probably isn't as much of a problem in practice.

Running an actual DaemonSet solves that problem. Could cron be put into a container?

**Comment by RenaudWasTaken** [comment]
Hello !

> DaemonSets are designed for long-running daemons, not for batch workloads. I still prefer we solve this in the CronJob API.

I'd like to contribute another use case. Running a job on each nodes to detect GPUs and label the node accordingly.

Currently for mixed clusters (GPUs and non GPUs nodes), which are pretty common, the NVIDIA GPU [device plugin](https://kubernetes.io/docs/concepts/cluster-administration/device-plugins/) which is a daemonset will be deployed on all nodes.

When it detects that a node does not have any GPU it will juste `select{}` forever. A better solution would be to either:
- Run a job on each node (and nodes that join) assigning GPU labels and have the GPU device plugin be gated by a node selector
- Have the GPU device plugin only restart on failure

I believe the second option was suggested by @resouer higher.
Are there any patterns I'm missing that might solve this (other than having cluster admins label their nodes manually)?

**Comment by rothgar** [comment]
Having a cron target of `@reboot` would be really handy for this. But need a way to do it on a per node basis. Would be much cleaner than long running daemons but may have complications with nodes where GPUs are hot pluggable.

If we do at `@reboot` can we also make sure we get `@teatime`?

**Comment by RenaudWasTaken** [comment]
> but may have complications with nodes where GPUs are hot pluggable.

FWIW we don't really support Hot pluggable GPUs or with very specific hardware.

**Comment by enisoc** [comment]
@RenaudWasTaken wrote:
> I'd like to contribute another use case. Running a job on each nodes to detect GPUs and label the node accordingly.

What about a DaemonSet with [node anti-affinity](https://kubernetes.io/docs/concepts/configuration/assign-pod-node/#affinity-and-anti-affinity) for any Node that already has been labeled (one way or another) with the relevant label key? Assuming DS re-syncs when Node labels change, the controller should remove the Pod after it does its job.

**Comment by RenaudWasTaken** [comment]
> What about a DaemonSet with node anti-affinity for any Node that already has been labeled (one way or another) with the relevant label key? Assuming DS re-syncs when Node labels change, the controller should remove the Pod after it does its job.

I think we can do that for now, but that sounds like a side effect that I'm not sure we can rely on.
Anyways I'll be testing that :)

**Comment by bgrant0607** [comment]
I can't imagine we'll do this.

**Comment by stepin** [comment]
> What about a DaemonSet with node anti-affinity for any Node that already has been labeled (one way or another) with the relevant label key? Assuming DS re-syncs when Node labels change, the controller should remove the Pod after it does its job.

As I understand, if pod is already scheduled it will be not deleted automatically. This is why there is https://github.com/kubernetes-incubator/descheduler project to delete pods if during runtime conditions (like labels on nodes) changed.

**Comment by stepin** [comment]
Our use case: init k8s node in GKE. As GKE have autoscaling of nodes it's impossible to use 3d-party configuration systems like Ansible. Also, GKE don't have ability to pre-create VM image for k8s node with required fixes (it will help for for some static configuration cases but not for all).

Now we can create DaemonSet that will do some work and finish with command like `tail -f /dev/null` (do nothing forever). So, it's hack to create something like Job from DaemonSet. Ideal solution will be to be able to configure run of Job on all nodes (or subset of nodes using labels).

**Comment by seanturner83** [comment]
Is this really never going to be done? :(

**Comment by OJFord** [comment]
@seanturner83 It's happening in #64623 by allowing DSs to have a `restartPolicy: OnFailure`.

**Comment by kucerarichard** [comment]
Right now I elected to use a Job kind,  but hardwire the affinity to each node,  with the help of Perl HEREDOC to generate the yaml.    I can't say I'm unhappy to write Perl again...

**Comment by unfor19** [comment]
I needed this to pull images (since they're large) upon Node creation. This way, the Pods creation will be significantly faster.

This is how I achieved it - https://codefresh.io/kubernetes-tutorial/single-use-daemonset-pattern-pre-pulling-images-kubernetes/

### TL;DR
1. DaemonSet [spec.initContainers](https://kubernetes.io/docs/concepts/workloads/pods/init-containers/) attribute - list of containers with the required images
2. DaemonSet `spec.containers` attribute - use the image `gcr.io/google_containers/pause`
3. The DaemonSet will run only once, so it will pull relevant images, and then stop

**Full example**
```
apiVersion: apps/v1
kind: DaemonSet
metadata:
  name: auto-pull-images
  namespace: default
  labels:
    k8s-app: auto-pull-images
spec:
  selector:
    matchLabels:
      name: auto-pull-images
  template:
    metadata:
      labels:
        name: auto-pull-images
    spec:
      initContainers:
        - name: serverless-template-container
          image: unfor19/serverless-template
          resources:
            limits:
              cpu: 100m
              memory: 100Mi
            requests:
              cpu: 100m
              memory: 100Mi
      containers:
        - name: pause
          image: gcr.io/google_containers/pause
          resources:
            limits:
              cpu: 50m
              memory: 50Mi
            requests:
              cpu: 50m
              memory: 50Mi
```

**Note** Each `auto-pull-images` Pod takes up to **1MB** on each Node, it's almost nothing, but just thought you should be aware of it

**Comment by dexterhu** [comment]
This looks like a pain point for many applications. We need a long term solution.

**Comment by dexterhu** [comment]
https://github.com/kubernetes/kubernetes/issues/24725 is not happening. Looks it is industry-wide pain point.

**Comment by adyanth** [comment]
Is there anything better than this?
```yaml
apiVersion: batch/v1
kind: CronJob
spec:
  jobTemplate:
    spec:
      completions: 5 # Number of nodes
      parallelism: 5 # Number of nodes
      template:
        metadata:
          labels:
            app.kubernetes.io/name: daemon-cron
        spec:
          affinity:
            podAntiAffinity:
              requiredDuringSchedulingIgnoredDuringExecution:
                - labelSelector:
                    matchExpressions:
                      - key: app.kubernetes.io/name
                        operator: In
                        values:
                          - daemon-cron
                  topologyKey: kubernetes.io/hostname
...
```

**Comment by david-nano** [comment]
External CRD provide support for this: https://openkruise.io/docs/user-manuals/advancedcronjob/
